modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java

```diff
@@ -575,32 +581,55 @@ public Weight createNormalizedWeight(Query query, boolean needsScores) throws IO
         }
     }
 
-    private static PercolateQuery.QueryStore createStore(PercolatorFieldMapper.FieldType fieldType,
-                                                         QueryShardContext context,
-                                                         boolean mapUnmappedFieldsAsString) {
+    static PercolateQuery.QueryStore createStore(MappedFieldType queryBuilderFieldType,
+                                                 QueryShardContext context,
+                                                 boolean mapUnmappedFieldsAsString) {
+        Version indexVersion = context.indexVersionCreated();
+        NamedWriteableRegistry registry = context.getWriteableRegistry();
         return ctx -> {
-            LeafReader leafReader = ctx.reader();
-            BinaryDocValues binaryDocValues = leafReader.getBinaryDocValues(fieldType.queryBuilderField.name());
-            if (binaryDocValues == null) {
-                return docId -> null;
-            }
-
-            return docId -> {
-                if (binaryDocValues.advanceExact(docId)) {
-                    BytesRef qbSource = binaryDocValues.binaryValue();
-                    if (qbSource.length > 0) {
-                        XContent xContent = PercolatorFieldMapper.QUERY_BUILDER_CONTENT_TYPE.xContent();
-                        try (XContentParser sourceParser = xContent.createParser(context.getXContentRegistry(), qbSource.bytes,
-                                qbSource.offset, qbSource.length)) {
-                            return parseQuery(context, mapUnmappedFieldsAsString, sourceParser);
+            if (indexVersion.onOrAfter(Version.V_6_0_0_beta1)) {
+                SortedBinaryDocValues values = context.getForField(queryBuilderFieldType).load(ctx).getBytesValues();
+                return docId -> {
+                    if (values.advanceExact(docId)) {
+                        BytesRef qbSource = values.nextValue();
```

Comment:
So this change is a side effect of making the the query builder field inside the percolator field accessible via doc values fields. I needed that in order to make the query builder bwc qa test to work. 

When field values via doc values fields it uses `ScriptDocValues` to access doc values. (DocValueFieldsFetchSubPhase.java line 88) This uses a special encoding to deal with multi values binary values.